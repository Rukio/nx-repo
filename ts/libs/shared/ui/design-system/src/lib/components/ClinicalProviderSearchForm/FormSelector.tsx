import { Button, ButtonGroup } from '../..';
import { makeSxStyles } from '../../utils/makeSxStyles';
import { ClinicalProviderSearchFormType } from './index';
import CLINICAL_PROVIDER_SEARCH_FORM_TEST_IDS from './testIds';

export interface FormSelectorProps {
  searchForms: ClinicalProviderSearchFormType[];
  selectedForm: ClinicalProviderSearchFormType;
  onSelect: (form: ClinicalProviderSearchFormType) => void;
}

const makeStyles = () =>
  makeSxStyles({
    buttonGroup: {
      marginBottom: 3,
    },
    buttonSelected: (theme) => ({
      background: theme.palette.primary.main,
      color: 'white',
      '&:hover': {
        background: theme.palette.primary.main,
      },
    }),
  });

const FormSelector: React.FC<FormSelectorProps> = ({
  searchForms,
  selectedForm,
  onSelect,
}) => {
  const styles = makeStyles();

  const isButtonSelected = (formName: ClinicalProviderSearchFormType) =>
    selectedForm === formName;

  const ButtonLabels = {
    [ClinicalProviderSearchFormType.Name]: 'Name',
    [ClinicalProviderSearchFormType.Practice]: 'Practice',
    [ClinicalProviderSearchFormType.Phone]: 'Phone',
    [ClinicalProviderSearchFormType.Pharmacy]: 'Pharmacy',
  };

  return (
    <ButtonGroup color="primary" fullWidth={true} sx={styles.buttonGroup}>
      {searchForms.map((formName) => (
        <Button
          key={`${CLINICAL_PROVIDER_SEARCH_FORM_TEST_IDS.FORM_SELECT_BUTTON}-${formName}`}
          data-testid={`${CLINICAL_PROVIDER_SEARCH_FORM_TEST_IDS.FORM_SELECT_BUTTON}-${formName}`}
          size="large"
          sx={isButtonSelected(formName) ? styles.buttonSelected : null}
          onClick={() => onSelect(formName)}
        >
          By {ButtonLabels[formName]}
        </Button>
      ))}
    </ButtonGroup>
  );
};

export default FormSelector;
