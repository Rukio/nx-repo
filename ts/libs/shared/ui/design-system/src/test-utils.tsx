import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider, theme } from './lib';
import type { Config as UserEventSetupConfig } from '@testing-library/user-event/setup/setup';
import userEvent from '@testing-library/user-event';

type CustomRenderOptions = {
  renderOptions?: Omit<RenderOptions, 'wrapper'>;
  userEventOptions?: UserEventSetupConfig;
};

const WrappedProviders: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement,
  options?: CustomRenderOptions
) => {
  return {
    user: userEvent.setup(options?.userEventOptions),
    ...render(ui, { wrapper: WrappedProviders, ...options?.renderOptions }),
  };
};

export * from '@testing-library/react';
export { customRender as render, userEvent };
