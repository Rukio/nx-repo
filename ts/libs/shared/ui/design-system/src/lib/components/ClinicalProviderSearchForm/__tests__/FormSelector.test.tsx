import { render, screen } from '../../../../test-utils';
import { Box } from '../../..';
import CLINICAL_PROVIDER_SEARCH_FORM_TEST_IDS from '../testIds';

import FormSelector, { FormSelectorProps } from '../FormSelector';
import { ClinicalProviderSearchFormType } from '../index';

const setup = ({ searchForms, selectedForm, onSelect }: FormSelectorProps) => {
  return render(
    <Box sx={{ width: '700px' }}>
      <FormSelector
        searchForms={searchForms}
        selectedForm={selectedForm}
        onSelect={onSelect}
      />
    </Box>
  );
};

const getButtonTestIdFor = (formName: string) =>
  `${CLINICAL_PROVIDER_SEARCH_FORM_TEST_IDS.FORM_SELECT_BUTTON}-${formName}`;

describe('ClinicalProviderSearchForm - Form Selector', () => {
  it('should render all form options correctly with test ids', () => {
    const mockOnSelect = jest.fn();
    const { asFragment } = setup({
      searchForms: [
        ClinicalProviderSearchFormType.Name,
        ClinicalProviderSearchFormType.Practice,
        ClinicalProviderSearchFormType.Phone,
      ],
      selectedForm: ClinicalProviderSearchFormType.Practice,
      onSelect: mockOnSelect,
    });
    expect(
      screen.getByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Name)
      )
    ).toHaveTextContent('By Name');
    expect(
      screen.getByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Practice)
      )
    ).toHaveTextContent('By Practice');
    expect(
      screen.getByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Phone)
      )
    ).toHaveTextContent('By Phone');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render subset of form options', () => {
    const mockOnSelect = jest.fn();
    setup({
      searchForms: [
        ClinicalProviderSearchFormType.Name,
        ClinicalProviderSearchFormType.Practice,
      ],
      selectedForm: ClinicalProviderSearchFormType.Name,
      onSelect: mockOnSelect,
    });
    expect(
      screen.getByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Name)
      )
    ).toHaveTextContent('By Name');
    expect(
      screen.getByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Practice)
      )
    ).toHaveTextContent('By Practice');
    expect(
      screen.queryByTestId(
        getButtonTestIdFor(ClinicalProviderSearchFormType.Phone)
      )
    ).not.toBeInTheDocument();
  });

  it('should notify which form selected', async () => {
    const mockOnSelect = jest.fn();
    const { user } = setup({
      searchForms: [
        ClinicalProviderSearchFormType.Name,
        ClinicalProviderSearchFormType.Practice,
        ClinicalProviderSearchFormType.Phone,
      ],
      selectedForm: ClinicalProviderSearchFormType.Phone,
      onSelect: mockOnSelect,
    });

    const practiceFormSelectButton = screen.getByTestId(
      getButtonTestIdFor(ClinicalProviderSearchFormType.Practice)
    );
    await user.click(practiceFormSelectButton);
    expect(mockOnSelect).toHaveBeenCalledWith(
      ClinicalProviderSearchFormType.Practice
    );
  });
});
