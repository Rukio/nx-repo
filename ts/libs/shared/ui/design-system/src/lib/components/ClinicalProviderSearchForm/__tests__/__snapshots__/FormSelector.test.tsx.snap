// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClinicalProviderSearchForm - Form Selector should render all form options correctly with test ids 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 700px;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 6px;
  box-shadow: none;
  width: 100%;
  margin-bottom: 24px;
}

.emotion-1 .MuiButtonGroup-grouped {
  min-width: 40px;
}

.emotion-1 .MuiButtonGroup-grouped:not(:first-of-type) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  margin-left: -1px;
}

.emotion-1 .MuiButtonGroup-grouped:not(:last-of-type) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-right-color: transparent;
}

.emotion-1 .MuiButtonGroup-grouped:not(:last-of-type):hover {
  border-right-color: currentColor;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-weight: 600;
  font-size: 0.9375rem;
  line-height: 1.714;
  text-transform: none;
  letter-spacing: 0.025rem;
  font-family: Open Sans,Segoe UI,Tahoma,sans-serif;
  min-width: 64px;
  padding: 7px 21px;
  border-radius: 6px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border: 1px solid rgba(0, 116, 184, 0.5);
  color: #0074B8;
  width: 100%;
  box-shadow: none;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(0, 116, 184, 0.04);
  border: 1px solid #0074B8;
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  color: rgba(17, 27, 34, 0.26);
  border: 1px solid rgba(17, 27, 34, 0.12);
}

.emotion-2:hover {
  box-shadow: none;
}

.emotion-2.Mui-focusVisible {
  box-shadow: none;
}

.emotion-2:active {
  box-shadow: none;
}

.emotion-2.Mui-disabled {
  box-shadow: none;
}

.emotion-3 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-weight: 600;
  font-size: 0.9375rem;
  line-height: 1.714;
  text-transform: none;
  letter-spacing: 0.025rem;
  font-family: Open Sans,Segoe UI,Tahoma,sans-serif;
  min-width: 64px;
  padding: 7px 21px;
  border-radius: 6px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border: 1px solid rgba(0, 116, 184, 0.5);
  color: #0074B8;
  width: 100%;
  box-shadow: none;
  background: #0074B8;
  color: white;
}

.emotion-4::-moz-focus-inner {
  border-style: none;
}

.emotion-4.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-4 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-4:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: rgba(0, 116, 184, 0.04);
  border: 1px solid #0074B8;
}

@media (hover: none) {
  .emotion-4:hover {
    background-color: transparent;
  }
}

.emotion-4.Mui-disabled {
  color: rgba(17, 27, 34, 0.26);
  border: 1px solid rgba(17, 27, 34, 0.12);
}

.emotion-4:hover {
  box-shadow: none;
}

.emotion-4.Mui-focusVisible {
  box-shadow: none;
}

.emotion-4:active {
  box-shadow: none;
}

.emotion-4.Mui-disabled {
  box-shadow: none;
}

.emotion-4:hover {
  background: #0074B8;
}

<div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiButtonGroup-root MuiButtonGroup-outlined MuiButtonGroup-fullWidth MuiButtonGroup-disableElevation emotion-1"
      role="group"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary emotion-2"
        data-testid="form-select-button-name"
        tabindex="0"
        type="button"
      >
        By Name
        <span
          class="MuiTouchRipple-root emotion-3"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary emotion-4"
        data-testid="form-select-button-practice"
        tabindex="0"
        type="button"
      >
        By Practice
        <span
          class="MuiTouchRipple-root emotion-3"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary emotion-2"
        data-testid="form-select-button-phone"
        tabindex="0"
        type="button"
      >
        By Phone
        <span
          class="MuiTouchRipple-root emotion-3"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
